posts/2020-01-30-Rust-1.41.0.md

```diff
@@ -0,0 +1,212 @@
+is now ABI compatible with the C language's pointer (`T*`) types. So if you
+have an `extern "C"` Rust function, called from C, your Rust function can now
+use `Box<T>`, for some specific `T`, while using `T*` in C for the
+corresponding function. As an example, on the C side you may have:
+
+```c
+// C header */
+
+// Returns ownership to the caller.
+struct Foo* foo_new(void);
+
+// Takes ownership from the caller; no-op when invoked with NULL.
+void foo_delete(struct Foo*);
+```
+
+...while on the Rust side, you would have:
+
+```rust
+#[repr(C)]
+pub struct Foo;
+
+#[no_mangle]
+pub extern "C" fn foo_new() -> Box<Foo> {
+    Box::new(Foo)
+}
+
+// The possibility of NULL is represented with the `Option<_>`.
+#[no_mangle]
+pub extern "C" fn foo_delete(_: Option<Box<Foo>>) {}
+```
+
+Note however that while `Box<T>` and `T*` have the same representation and ABI,
+a `Box<T>` must still be non-null, aligned, and ready for deallocation by the
+global allocator. To ensure this, it is best to only use `Box`es originating
+from the global allocator.
+
+**Important:** At least at present, you should avoid using `Box<T>` types for
+functions that are defined in C but invoked from Rust. In those cases, you
+should directly mirror the C types as closely as possible. Using types like
+`Box<T>` where the C definition is just using `T*` can lead to undefined
+behavior.
+
+To read more, [consult the documentation for `Box<T>`][box_docs].
+
+### Library changes
+
+[`Result::map_or`]: https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or
+[`Result::map_or_else`]: https://doc.rust-lang.org/std/result/enum.Result.html#method.map_or_else
+[`Option::map_or`]: https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or
+[`Option::map_or_else`]: https://doc.rust-lang.org/std/option/enum.Option.html#method.map_or_else
+[`std::rc::Weak::weak_count`]: https://doc.rust-lang.org/std/rc/struct.Weak.html#method.weak_count
+[`std::rc::Weak::strong_count`]: https://doc.rust-lang.org/std/rc/struct.Weak.html#method.strong_count
+[`std::sync::Weak::weak_count`]: https://doc.rust-lang.org/std/sync/struct.Weak.html#method.weak_count
+[`std::sync::Weak::strong_count`]: https://doc.rust-lang.org/std/sync/struct.Weak.html#method.strong_count
+[pr_66277]: https://github.com/rust-lang/rust/pull/66277
+[pr_65013]: https://github.com/rust-lang/rust/pull/65013
+
+In Rust 1.41.0, we've made the following additions to the standard library:
+
+- The [`Result::map_or`] and [`Result::map_or_else`] methods were stabilized.
+
+  Similar to [`Option::map_or`] and [`Option::map_or_else`], these methods are
+  shorthands for the `.map(|val| process(val)).unwrap_or(default)` pattern.
+
+- [`NonZero*` numerics now implement `From<NonZero*>` if it's a smaller integer
+  width.][pr_66277] For example, `NonZeroU16` now implements `From<NonZeroU8>`.
+
+- The `weak_count` and `strong_count` methods were stabilized.
```

Comment:
These were already stable on `Arc`/`Rc`, only the ones on `Weak` are new, so maybe:

```suggestion
- The `weak_count` and `strong_count` methods on `Weak` pointers were stabilized.
```